processor/src/operations/sys_ops.rs

```diff
@@ -11,9 +11,146 @@ impl Process {
+    /// Returns an error if the stack is empty.
+    pub(super) fn op_fmpadd(&mut self) -> Result<(), ExecutionError> {
+        self.stack.check_depth(1, "FMPADD")?;
+
+        let offset = self.stack.get(0);
+        let fmp = self.system.fmp();
+
+        self.stack.set(0, fmp + offset);
+        self.stack.copy_state(1);
+
+        Ok(())
+    }
+
+    /// Pops an element off the stack and adds it to the current value of `fmp` register.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The stack is empty.
+    /// * New value of `fmp` register is greater than or equal to 2^32.
+    pub(super) fn op_fmpupdate(&mut self) -> Result<(), ExecutionError> {
+        self.stack.check_depth(1, "FMPUPDATE")?;
+
+        let offset = self.stack.get(0);
+        let fmp = self.system.fmp();
+
+        let new_fmp = fmp + offset;
+        if new_fmp.as_int() > u32::MAX as u64 {
+            return Err(ExecutionError::InvalidFmpValue(fmp, new_fmp));
+        }
+
+        self.system.set_fmp(new_fmp);
+        self.stack.shift_left(1);
+
+        Ok(())
+    }
+}
+
+// TESTS
+// ================================================================================================
+
+#[cfg(test)]
+mod tests {
+    use super::{
+        super::{init_stack_with, Operation},
+        Felt, FieldElement, Process,
+    };
+
+    #[test]
+    fn op_fmpupdate() {
+        let mut process = Process::new_dummy();
+
+        // initial value of fmp register should be zero
+        assert_eq!(Felt::ZERO, process.system.fmp());
+
+        // increment fmp register
+        process.execute_op(Operation::Push(Felt::new(2))).unwrap();
+        process.execute_op(Operation::FmpUpdate).unwrap();
+        assert_eq!(Felt::new(2), process.system.fmp());
+
+        // increment fmp register again
+        process.execute_op(Operation::Push(Felt::new(3))).unwrap();
+        process.execute_op(Operation::FmpUpdate).unwrap();
+        assert_eq!(Felt::new(5), process.system.fmp());
+
+        // decrement fmp register
+        process.execute_op(Operation::Push(-Felt::new(3))).unwrap();
+        process.execute_op(Operation::FmpUpdate).unwrap();
+        assert_eq!(Felt::new(2), process.system.fmp());
+
+        // decrementing beyond zero should be an error
+        process.execute_op(Operation::Push(-Felt::new(3))).unwrap();
+        assert!(process.execute_op(Operation::FmpUpdate).is_err());
+
+        // going up to u32::MAX should be OK
+        let mut process = Process::new_dummy();
+        process
+            .execute_op(Operation::Push(Felt::new(u32::MAX as u64)))
+            .unwrap();
+        process.execute_op(Operation::FmpUpdate).unwrap();
+        assert_eq!(Felt::new(u32::MAX as u64), process.system.fmp());
+
+        // but going beyond that should be an error
+        process
+            .execute_op(Operation::Push(Felt::new(u32::MAX as u64 + 1)))
+            .unwrap();
+        assert!(process.execute_op(Operation::FmpUpdate).is_err());
```

Comment:
Just a note that since the process wasn't reset in between this seems to be checking for an error at 2 * u32::MAX + 1, which may not have been the intention

Reply:
Great catch again! fixed!